java.library.path should be configurable when launching the application's native image

jaikiran opened the pull request:
Fixes https://github.com/quarkusio/quarkus/issues/5880

The commit here ensures that the `java.library.path` is configurable at runtime when launching the application's native image (by passing `-Djava.library.path` to point to additional paths). This also includes a test case which reproduces the issue and verifies the fix.

jaikiran committed:
issue-5880 java.library.path should be configurable when launching the application's native image

jaikiran commented:
The CI failed with:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:native-image (default) on project acme: Failed to generate a native image: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.RuntimeException: Cannot find the `native-image` in the GRAALVM_HOME, JAVA_HOME and System PATH. Install it using `gu install native-image`
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.getNativeImageExecutable(NativeImageBuildStep.java:336)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:131)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:921)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
[ERROR] 	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1535)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1426)
[ERROR] 	at java.lang.Thread.run(Thread.java:748)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
[ERROR] -> [Help 1]
```
Looks like it isn't setup with native image generation for Windows? I'll update the PR to run the test only against *nix.

cescoffier commented on integration-tests/maven/src/test/resources/projects/native-image-app/src/main/resources/META-INF/resources/index.html:
```diff
@@ -0,0 +1,153 @@
+<!DOCTYPE html>
```
Could be removed.

cescoffier commented on integration-tests/maven/src/test/resources/projects/native-image-app/src/main/java/org/acme/MyApplication.java:
```diff
@@ -0,0 +1,9 @@
+package org.acme;
```
Could be removed.

cescoffier commented on integration-tests/maven/src/test/resources/projects/native-image-app/pom.xml:
```diff
@@ -0,0 +1,85 @@
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-resteasy</artifactId>
+    </dependency>
+    <dependency>
```
Could be removed.

jaikiran replied:
Hello @cescoffier, The test application needs this since it exposes a REST endpoint to invoke upon from the test. I've retained this dependency. Let me know if I missed something.

cescoffier commented on integration-tests/maven/src/test/resources/projects/native-image-app/pom.xml:
```diff
@@ -0,0 +1,85 @@
+      <artifactId>quarkus-junit5</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
```
Could be removed.

cescoffier commented on integration-tests/maven/src/test/resources/projects/native-image-app/src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+# Configuration file
```
Could be removed.

cescoffier reviewed (dismissed):
I believe we could clean up the test a bit. Many files are not really required.

machi1990 commented:
> I believe we could clean up the test a bit. Many files are not really required.

Maybe we can use the existing `project/classic` setup and pass in the necessary configurations for this particular test case?

jaikiran commented:
> Maybe we can use the existing project/classic setup and pass in the necessary configurations for this particular test case?

Hello @machi1990, I didn't want to interfere with what that test project and its testcases were doing by adding this slightly unrelated test case into it, especially since this test project requires to enable native SSL, which may not be a good idea to introduce into the "classic" project.

jaikiran commented:
Hello @machi1990 and @cescoffier, thank you both for the reviews. I've updated this PR to incorporate your reviews. I've also adjusted the test case a bit to handle cases where the `native-image` command isn't available, by skipping this test in those setups (for example the couple of setups of our CI where this test case was failing).

machi1990 commented:
> > Maybe we can use the existing project/classic setup and pass in the necessary configurations for this particular test case?
> 
> Hello @machi1990, I didn't want to interfere with what that test project and its testcases were doing by adding this slightly unrelated test case into it, especially since this test project requires to enable native SSL, which may not be a good idea to introduce into the "classic" project.

I see the reasoning. I was thinking of modifying the configuration file programatically as done in other `DevMojo` tests.

gsmet commented:
Looks OK to me, I merged it, thanks!
